### PR TITLE
adding app gateway module

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
Adding support for Azure Application Gateway
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
azure_rm_appgw

##### ANSIBLE VERSION
2.5

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 1]
```diff
@@ -0,0 +1,1085 @@
...
+        name: rule1
+'''
+
+RETURN = '''
+id:
+    description:
+        - Resource ID.
+    returned: always
+    type: str
+    sample: id
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: add sample?

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 2]
```diff
@@ -0,0 +1,1085 @@
...
+                setattr(self, key, kwargs[key])
+            elif kwargs[key] is not None:
+                if key == "id":
+                    self.parameters["id"] = kwargs[key]
+                elif key == "location":
+                    self.parameters["location"] = kwargs[key]
+                elif key == "sku":
+                    ev = kwargs[key]
+                    if 'name' in ev:
+                        if ev['name'] == 'standard_small':
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: why not directly define choices of name as "Standard_Small"?

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 3]
```diff
@@ -0,0 +1,1085 @@
...
+                    self.parameters["enable_http2"] = kwargs[key]
+                elif key == "resource_guid":
+                    self.parameters["resource_guid"] = kwargs[key]
+                elif key == "etag":
+                    self.parameters["etag"] = kwargs[key]
+
+        old_response = None
+        response = None
+
+        self.mgmt_client = self.get_mgmt_svc_client(NetworkManagementClient,
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: why not self._network_client  defined in azure_rm_common?

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py (hunk 2 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: just to follow python case standard. that's current idea.

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py (hunk 3 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: i will use

File: test/integration/targets/azure_rm_appgw/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        http_listener: sample_http_listener
+        name: rule1
+  check_mode: yes
+  register: output
+- name: Assert the resource instance is well created
+  assert:
+    that:
+      - output.changed
+
+- name: dump
```
Comment (yaweiw) [inline test/integration/targets/azure_rm_appgw/tasks/main.yml]: Capital 'D'?

Comment (zikalino) [reply test/integration/targets/azure_rm_appgw/tasks/main.yml]: ah this needs to be removed entirely

File: test/integration/targets/azure_rm_appgw/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  debug:
+    var: output
+
+- name: Create instance of Application Gateway
```
Comment (yungezz) [inline test/integration/targets/azure_rm_appgw/tasks/main.yml]: add a test to cover update existing appgateway?

Comment (zikalino) [reply test/integration/targets/azure_rm_appgw/tasks/main.yml]: still have to add idempotency check for existing items, but will do it by the end of today

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 6]
```diff
@@ -0,0 +1,805 @@
...
+                    - 'app_gw_ssl_policy20170401'
+                    - 'app_gw_ssl_policy20170401_s'
+            cipher_suites:
+                description:
+                    - List of SSL cipher suites to be enabled in the specified order to application gateway.
+            min_protocol_version:
+                description:
+                    - Minimum version of Ssl protocol to be supported on application gateway.
+                choices:
+                    - 'tl_sv1_0'
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: Why the extra `_` in these values vs what the REST API uses?

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 7]
```diff
@@ -0,0 +1,805 @@
...
+    gateway_ip_configurations:
+        description:
+            - List of subnets used by the application gateway.
+        suboptions:
+            subnet:
+                description:
+                    - Reference of the subnet resource. A subnet from where application gateway gets its private address.
+            name:
+                description:
+                    - Name of the resource that is unique within a resource group. This name can be used to access the resource.
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: s/Name of the resource/Name for the gateway IP configuration/ (if I'm understanding this properly)

**Comment by kyliel** [comment]
@zikalino , is it ready for review? Thanks.

**Comment by zikalino** [comment]
@kyliel yes, it was already checked by @nitzmahone last week, just had comments about using snake_to_camel wherever possible, but need full review

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 8]
```diff
@@ -0,0 +1,792 @@
...
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_appgw
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: should this be called `azure_rm_appgateway` instead? Have we used aliases in any other module?

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 9]
```diff
@@ -0,0 +1,792 @@
...
+version_added: "2.6"
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: 2.7

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 10]
```diff
@@ -0,0 +1,792 @@
...
+            location=dict(
+                type='str'
+            ),
+            sku=dict(
+                type='dict'
+            ),
+            ssl_policy=dict(
+                type='dict'
+            ),
+            gateway_ip_configurations=dict(
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: You can define a subspec for dictionaries, useful for things like the `password` field in `ssl_certificates` where you don't want to log the value.

File: test/integration/targets/azure_rm_appgw/aliases [hunk 11]
```diff
@@ -0,0 +1,3 @@
+cloud/azure
+destructive
+posix/ci/cloud/group2/azure
```
Comment (jborean93) [inline test/integration/targets/azure_rm_appgw/aliases]: The alias was just recently changed, see https://github.com/ansible/ansible/blob/devel/test/integration/targets/azure_rm_loadbalancer/aliases for the new naming standard for this field https://github.com/ansible/ansible/blob/devel/test/integration/targets/azure_rm_loadbalancer/aliases

File: lib/ansible/modules/cloud/azure/azure_rm_appgw.py [hunk 12]
```diff
@@ -0,0 +1,792 @@
...
+    newd = {}
+    for item in new:
+        name = item['name']
+        newd[name] = item
+
+    newd = dict_merge(oldd, newd)
+    return newd == oldd
+
+
+def dict_merge(a, b):
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appgw.py]: Was this recently added to another module_util, I seem to recall it being used somewhere else.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/azure/azure_rm_appgateway.py [hunk 13]
```diff
@@ -0,0 +1,789 @@
...
+            capacity:
+                description:
+                    - Capacity (instance count) of an application gateway.
+    ssl_policy:
+        description:
+            - SSL policy of the application gateway resource.
+        suboptions:
+            disabled_ssl_protocols:
+                description:
+                    - List of SSL protocols to be disabled on application gateway.
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_appgateway.py:64]: format/choices?

File: lib/ansible/modules/cloud/azure/azure_rm_appgateway.py [hunk 14]
```diff
@@ -0,0 +1,789 @@
...
+            policy_name:
+                description:
+                    - Name of Ssl C(predefined) policy.
+                choices:
+                    - 'ssl_policy20150501'
+                    - 'ssl_policy20170401'
+                    - 'ssl_policy20170401_s'
+            cipher_suites:
+                description:
+                    - List of SSL cipher suites to be enabled in the specified order to application gateway.
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_appgateway.py:84]: format/choices?

**Comment by Fred-sun** [comment]
@yaweiw @yungezz @jborean93 @nitzmahone The PR has ready for reviewed, Could you help review it? Thanks!

**Comment by Fred-sun** [comment]
@zikalino Could you help to confirm the PR ready for review or not? Thanks!

**Comment by Fred-sun** [comment]
@zikalino Could you help to confirm the PR ready for review or not? Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_appgateway.py [hunk 15]
```diff
@@ -0,0 +1,790 @@
...
+            name:
+                description:
+                    - Name of the resource that is unique within a resource group. This name can be used to access the resource.
+    authentication_certificates:
+        description:
+            - Authentication certificates of the application gateway resource.
+        suboptions:
+            data:
+                description:
+                    - Certificate public data.
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_appgateway.py]: + base64 pfx encoded (or whatever this one needs)

**Review by nitzmahone — CHANGES_REQUESTED** [review]
couple more docs things, then good to go

**Comment by zikalino** [comment]
@nitzmahone fixed these issues

**Review by nitzmahone — APPROVED** [review]

**Pull request merged**